src-built-in/components/UIAPI/reducers/linker.ts

```diff
@@ -0,0 +1,184 @@
+import { Channel, Channels, NameToId, Linker, LinkerAction } from '../types';
+
+declare var FSBL: any;
+declare var finsembleWindow: any;
+
+const initialState: Linker = {
+    channels: {},
+    nameToId: {},
+    isAccessibleLinker: true,
+    windowIdentifier: {},
+    processingRequest: false
+};
+
+// Effectful code to link/unlink the channel which will run outside the reducer function
+function linkChannel(channelName: string, isActive: boolean, windowIdentifier: object) {
+    return new Promise((res, rej) => {
+        const callback = (err: any, data: object) => {
+            if (err) return rej(err);
+            res(data);
+        };
+        if (!isActive) {
+            console.log("linking the window: ", windowIdentifier);
+            FSBL.Clients.LinkerClient.linkToChannel(channelName, windowIdentifier, callback);
+        } else {
+            FSBL.Clients.LinkerClient.unlinkFromChannel(channelName, windowIdentifier, callback);
+        }
+    });
+}
+
+function initializeLinker() {
+    finsembleWindow.addEventListener("blurred", () => {
+        finsembleWindow.hide();
+    });
+    finsembleWindow.addEventListener("shown", () => {
+		finsembleWindow.focus();
+    });
+    
+    let nextChannelId: number = 0;
+    const linkerInitState: Linker = Object.assign({}, initialState);
+    const initialChannels: Channels = {};
+    const initialNametoId: NameToId = {};
+
+    return new Promise((res, rej) => {
+        FSBL.Clients.LinkerClient.getAllChannels().forEach((channel: Channel) => {
+            initialChannels[nextChannelId] = {
+                id: nextChannelId,
+                name: channel.name,
+                color: channel.color,
+                active: false
+            };
+            initialNametoId[channel.name] = nextChannelId;
+            nextChannelId += 1;
+        });
+        linkerInitState.channels = initialChannels;
+        linkerInitState.nameToId = initialNametoId;
+
+        FSBL.Clients.RouterClient.addResponder("Finsemble.LinkerWindow.SetActiveChannels", function (err: any, msg: any) {
+            if (err) {
+                return rej(`Failed to add Finsemble.LinkerWindow.SetActiveChannels Responder: ${err}`);
+            }
+            const activeChannels = msg.data.channels;
+
+            const activeChannelIds: any[number] = [];
+            activeChannels.forEach((channel: Channel) => {
+                activeChannelIds.push(linkerInitState.nameToId[channel.name]);
+            });
+            const updatedChannels: Channels = Object.assign({}, linkerInitState.channels);
+            activeChannelIds.forEach((channelId: number) => {
+                updatedChannels[channelId].active = true;
+            });
+
+            linkerInitState.windowIdentifier = msg.data.windowIdentifier;
+            FSBL.Clients.Logger.system.log("toggle Linker window");
+            msg.sendQueryResponse(null, {});
+            store.dispatch(updateActiveChannels(msg.data));
```

Comment:
When we initialize the component we would send 2 updates to the linker state with the same payload. First is through `store.dispatch`, second is through `res(newLinkerState)`. The store will update twice with the same content, so it wouldn't cause any issues.

We need the `store.dispatch` there because we're listening on any channel updates when user switches the linker window for different components so that we can dispatch an action to update the active channels. `res(newLinkerState)` wouldn't be invoked after the initial linker initialization. 

Please let me know if this is still unclear, I can explain in person.